pkg/clients/s3/bucketclients/CORSConfig.go

```diff
@@ -0,0 +1,138 @@
+limitations under the License.
+*/
+
+package bucketclients
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go-v2/aws/awserr"
+	awss3 "github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+
+	"github.com/crossplane/provider-aws/apis/s3/v1beta1"
+	aws "github.com/crossplane/provider-aws/pkg/clients"
+	"github.com/crossplane/provider-aws/pkg/clients/s3"
+)
+
+// CORSConfigurationClient is the client for API methods and reconciling the CORSConfiguration
+type CORSConfigurationClient struct {
+	config *v1beta1.CORSConfiguration
+	client s3.BucketClient
+}
+
+// NewCORSConfigurationClient creates the client for CORS Configuration
+func NewCORSConfigurationClient(bucket *v1beta1.Bucket, client s3.BucketClient) *CORSConfigurationClient {
+	return &CORSConfigurationClient{config: bucket.Spec.Parameters.CORSConfiguration, client: client}
+}
+
+func (in *CORSConfigurationClient) corsNotFound(err error) bool {
+	if s3Err, ok := err.(awserr.Error); ok && s3Err.Code() == "NoSuchCORSConfiguration" && in.config == nil {
+		return true
+	}
+	return false
+}
+
+func sameRule(local *v1beta1.CORSRule, external *awss3.CORSRule) bool {
+	if !cmp.Equal(local.AllowedHeaders, external.AllowedHeaders) {
+		return false
+	}
+	if !cmp.Equal(local.AllowedMethods, external.AllowedMethods) {
+		return false
+	}
+	if !cmp.Equal(local.AllowedOrigins, external.AllowedOrigins) {
+		return false
+	}
+	if !cmp.Equal(local.ExposeHeaders, external.ExposeHeaders) {
+		return false
+	}
+	if !cmp.Equal(local.MaxAgeSeconds, external.MaxAgeSeconds) {
+		return false
+	}
+	return true
+}
+
+func compareCORS(local *v1beta1.CORSConfiguration, external []awss3.CORSRule) ResourceStatus {
+	if local == nil && external != nil {
+		return NeedsDeletion
+	} else if local == nil && len(external) == 0 {
+		return Updated
+	}
+
+	if len(local.CORSRules) != len(external) {
+		return NeedsUpdate
+	}
+
+	for i := range local.CORSRules {
+		outputRule := external[i]
+		if !sameRule(&local.CORSRules[i], &outputRule) {
+			return NeedsUpdate
+		}
+	}
+	return Updated
+}
+
+// Observe checks if the resource exists and if it matches the local configuration
+func (in *CORSConfigurationClient) Observe(ctx context.Context, bucket *v1beta1.Bucket) (ResourceStatus, error) {
+	conf, err := in.client.GetBucketCorsRequest(&awss3.GetBucketCorsInput{Bucket: aws.String(meta.GetExternalName(bucket))}).Send(ctx)
+	if err != nil && in.corsNotFound(err) {
+		return Updated, nil
+	} else if err != nil {
+		return NeedsUpdate, errors.Wrap(err, "cannot get bucket CORS configuration")
+	}
+
+	return compareCORS(in.config, conf.CORSRules), nil
+}
+
+// GeneratePutBucketCorsInput creates the input for the PutBucketCors request for the S3 Client
+func (in *CORSConfigurationClient) GeneratePutBucketCorsInput(name string) *awss3.PutBucketCorsInput {
```

Comment:
Sounds good, I would prefer to keep late init as a part of the object through, since it is part of the interface as well this makes it easier to consistently call the late-init function.